File: 2019m./uzd2./sprendimai/uzd2.py
class Contestant:
    def __init__(self, name, id, time):
        self.name = name
        self.id = id
        self.time = time

def getIndexById(id, arr):
    for i in range(len(arr)):
        if(arr[i].id == id):
            return i

def sortContestants(contestants):
    size = len(contestants)
    for i in range(size):
        for j in range(i + 1, size):
            if(contestants[i].time > contestants[j].time or (
                contestants[i].time == contestants[j].time
                and contestants[i].name > contestants[j].name)):
                contestants[j], contestants[i] = contestants[i], contestants[j]
    return contestants

def writeOutput(output, finishers, gender):
    output.write(gender + "\n")
    for i in range(len(finishers)):
        seconds = finishers[i].time % 60
        finishers[i].time -= seconds
        finishers[i].time /= 60
        minutes = int(finishers[i].time % 60)
        hours = int((finishers[i].time - minutes) / 60)
        output.write(str(finishers[i].id) + " " + finishers[i].name + " " * (20 - len(finishers[i].name)) + str(hours) + " " + str(minutes) + " " + str(seconds) + "\n") 

def main():
    data = open("U2.txt", "r")
    output = open("U2rez.txt", "w")

    contestants = []

    for i in range(int(next(data))):
        contestantData = next(data)
        name = contestantData[:19]
        contestantData = contestantData[20:].split(" ")
        contestantData = [int(o) for o in contestantData]
        startTime = contestantData[1] * 3600 + contestantData[2] * 60 + contestantData[3]
        contestants.append(Contestant(name, contestantData[0], startTime))

    femaleFinishers = []
    maleFinishers = []

    for i in range(int(next(data))):
        contestantData = next(data).split(" ")
        contestantData = [int(o) for o in contestantData]
        contestantIndex = getIndexById(contestantData[0], contestants)
        startTime = contestants[contestantIndex].time
        endTime = contestantData[1] * 3600 + contestantData[2] * 60 + contestantData[3]

        if(contestantData[0] >= 200):
            penaltyTime = (20 - contestantData[4] - contestantData[5] 
                              - contestantData[6] - contestantData[7]) * 60
            finishTime = endTime - startTime + penaltyTime
            maleFinishers.append(Contestant(contestants[contestantIndex].name, contestantData[0], finishTime))
        else:
            penaltyTime = (10 - contestantData[4] - contestantData[5]) * 60
            finishTime = endTime - startTime + penaltyTime
            femaleFinishers.append(Contestant(contestants[contestantIndex].name, contestantData[0], finishTime))
    
    femaleFinishers = sortContestants(femaleFinishers)
    maleFinishers = sortContestants(maleFinishers)

    writeOutput(output, femaleFinishers, "Merginos")
    writeOutput(output, maleFinishers, "Vaikinai")

    data.close()
    output.close()

if __name__ == "__main__":
    main()

File: 2019m./uzd2./sprendimai/uzd2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

class Contestant {
public:
    string name;
    int id;
    int time;

    Contestant(const string& name, int id, int time) : name(name), id(id), time(time) {}
};

int getIndexById(int id, const vector<Contestant>& arr) {
    for (size_t i = 0; i < arr.size(); ++i) {
        if (arr[i].id == id) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

vector<Contestant> sortContestants(vector<Contestant> contestants) {
    size_t size = contestants.size();
    for (size_t i = 0; i < size; ++i) {
        for (size_t j = i + 1; j < size; ++j) {
            if (contestants[i].time > contestants[j].time ||
                (contestants[i].time == contestants[j].time &&
                 contestants[i].name > contestants[j].name)) {
                swap(contestants[i], contestants[j]);
            }
        }
    }
    return contestants;
}

void writeOutput(ofstream& output, const vector<Contestant>& finishers, const string& gender) {
    output << gender << "\n";
    for (const Contestant& contestant : finishers) {
        int seconds = contestant.time % 60;
        int minutes = (contestant.time / 60) % 60;
        int hours = contestant.time / 3600;
        output << contestant.id << " " << contestant.name << string(20 - contestant.name.length(), ' ')
               << hours << " " << minutes << " " << seconds << "\n";
    }
}

int main() {
    ifstream data("U2.txt");
    ofstream output("U2rez.txt");

    vector<Contestant> contestants;

    int numContestants;
    data >> numContestants;
    data.ignore();

    for (int i = 0; i < numContestants; ++i) {
        string contestantData;
        getline(data, contestantData);
        string name = contestantData.substr(0, 20);
        contestantData = contestantData.substr(20);
        stringstream dataStream(contestantData);
        int id, hours, minutes, seconds;
        dataStream >> id >> hours >> minutes >> seconds;
        int startTime = hours * 3600 + minutes * 60 + seconds;
        contestants.push_back(Contestant(name, id, startTime));
    }

    vector<Contestant> femaleFinishers;
    vector<Contestant> maleFinishers;

    int numFinishers;
    data >> numFinishers;
    data.ignore();

    for (int i = 0; i < numFinishers; ++i) {
        string contestantData;
        getline(data, contestantData);
        stringstream dataStream(contestantData);
        int id, hours, minutes, seconds;
        dataStream >> id >> hours >> minutes >> seconds;
        vector<int> shots;
        int value;
        while (dataStream >> value) {
            shots.push_back(value);
        }
        int contestantIndex = getIndexById(id, contestants);
        int startTime = contestants[contestantIndex].time;
        int endTime = hours * 3600 + minutes * 60 + seconds;

        if (id >= 200) {
            int penaltyTime = (20 - shots[0] - shots[1] - shots[2] - shots[3]) * 60;
            int finishTime = endTime - startTime + penaltyTime;
            maleFinishers.push_back(Contestant(contestants[contestantIndex].name, id, finishTime));
        } else {
            int penaltyTime = (10 - shots[0] - shots[1]) * 60;
            int finishTime = endTime - startTime + penaltyTime;
            femaleFinishers.push_back(Contestant(contestants[contestantIndex].name, id, finishTime));
        }
    }

    femaleFinishers = sortContestants(femaleFinishers);
    maleFinishers = sortContestants(maleFinishers);

    writeOutput(output, femaleFinishers, "Merginos");
    writeOutput(output, maleFinishers, "Vaikinai");

    data.close();
    output.close();

    return 0;
}
